internal/configsource/vaultconfigsource/session.go

```diff
@@ -0,0 +1,318 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vaultconfigsource
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/vault/api"
+	"go.opentelemetry.io/collector/config/experimental/configsource"
+	"go.uber.org/zap"
+
+	"github.com/signalfx/splunk-otel-collector/internal/configprovider"
+)
+
+var errInvalidPollInterval = errors.New("poll interval must be greater than zero")
+
+// Error wrapper types to help with testability
+type (
+	errClientRead    struct{ error }
+	errNilSecret     struct{ error }
+	errNilSecretData struct{ error }
+	errBadSelector   struct{ error }
+)
+
+// vaultSession implements the configsource.Session interface.
+type vaultSession struct {
+	logger *zap.Logger
+	client *api.Client
+	secret *api.Secret
+
+	doneCh chan struct{}
+
+	path string
+
+	pollInterval time.Duration
+}
+
+var _ configsource.Session = (*vaultSession)(nil)
+
+func (v *vaultSession) Retrieve(_ context.Context, selector string, _ interface{}) (configsource.Retrieved, error) {
+	// By default assume that watcher is not supported. The exception will be the first
+	// value read from the vault secret.
+	watchForUpdateFn := watcherNotSupported
+
+	if v.secret == nil {
+		if err := v.readSecret(); err != nil {
+			return nil, err
+		}
+
+		// The keys come all from the same secret so creating a watcher only for the
+		// first it is fine.
+		var err error
+		watchForUpdateFn, err = v.buildWatcherFn()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	value := traverseToKey(v.secret.Data, selector)
+	if value == nil {
+		return nil, &errBadSelector{fmt.Errorf("no value at path %q for key %q", v.path, selector)}
+	}
+
+	return configprovider.NewRetrieved(value, watchForUpdateFn), nil
+}
+
+func (v *vaultSession) RetrieveEnd(context.Context) error {
+	return nil
+}
+
+func (v *vaultSession) Close(context.Context) error {
+	close(v.doneCh)
+
+	// Vault doesn't have a close for its client, close is completed.
+	return nil
+}
+
+func newSession(client *api.Client, path string, logger *zap.Logger, pollInterval time.Duration) (*vaultSession, error) {
+	if pollInterval <= 0 {
+		return nil, errInvalidPollInterval
+	}
+
+	return &vaultSession{
+		logger:       logger,
+		client:       client,
+		path:         path,
+		pollInterval: pollInterval,
+		doneCh:       make(chan struct{}),
+	}, nil
+}
+
+func (v *vaultSession) readSecret() error {
```

Comment:
Is this function expected to leave a `nil` in `v.secret` in case of error?

Reply:
Yes, the expectation is that the caller will stop trying to retrieve data, since this an error that triggers configuration failure, but if called again it will try once more to retrieve the secret.

Reply:
Added a comment to make clear that this is the intended behavior.